plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/AbstractTestIcebergSmoke.java

```diff
@@ -1616,4 +1639,81 @@ public void testOptimizedMetadataQueries()
+    @Test
+    public void testIncorrectIcebergFileSizes()
+            throws Exception
+    {
+        // Create a table with a single insert
+        assertUpdate("CREATE TABLE test_iceberg_file_size (x BIGINT) WITH (format='PARQUET')");
+        assertUpdate("INSERT INTO test_iceberg_file_size VALUES (123), (456), (758)", 3);
+
+        // Get manifest file
+        MaterializedResult result = computeActual("SELECT path FROM \"test_iceberg_file_size$manifests\"");
+        assertEquals(result.getRowCount(), 1);
+        String manifestFile = (String) result.getOnlyValue();
+
+        // Read manifest file
+        Schema schema;
+        GenericData.Record entry = null;
+        try (DataFileReader<GenericData.Record> dataFileReader = new DataFileReader<>(new File(manifestFile), new GenericDatumReader<>())) {
+            schema = dataFileReader.getSchema();
+            int recordCount = 0;
+            while (dataFileReader.hasNext()) {
+                entry = dataFileReader.next();
+                recordCount++;
+            }
+            assertEquals(recordCount, 1);
+        }
+
+        // Alter data file entry to store incorrect file size
+        GenericData.Record dataFile = (GenericData.Record) entry.get("data_file");
+        long alteredValue = 50L;
+        assertNotEquals((long) dataFile.get("file_size_in_bytes"), alteredValue);
+        dataFile.put("file_size_in_bytes", alteredValue);
+        entry.put("data_file", dataFile);
+
+        // Replace the file through HDFS client. This is required for correct checksums.
+        HdfsEnvironment hdfsEnvironment = createTestHdfsEnvironment();
+        HdfsEnvironment.HdfsContext context = new HdfsContext(getSession().toConnectorSession(), "iceberg");
+        Path manifestFilePath = new Path(manifestFile);
+        FileSystem fs = hdfsEnvironment.getFileSystem(context, manifestFilePath);
+        fs.delete(manifestFilePath, false);
+        fs.create(manifestFilePath);
+
+        // Write altered metadata
+        try (DataFileWriter<GenericData.Record> dataFileWriter = new DataFileWriter<>(new GenericDatumWriter<>(schema))) {
+            dataFileWriter.create(schema, new File(manifestFile));
+            dataFileWriter.append(entry);
+        }
+
+        // Ignoring Iceberg provided file size makes the query succeed
+        Session session = Session.builder(getSession())
+                .setCatalogSessionProperty("iceberg", "use_file_size_from_metadata", "false")
+                .build();
+        assertQuery(session, "SELECT * FROM test_iceberg_file_size", "VALUES (123), (456), (758)");
+
+        // Using Iceberg provided file size fails the query
+        assertThatThrownBy(() -> computeActual(getSession(), "SELECT * FROM test_iceberg_file_size"))
+                .hasMessageMatching(format("Error reading tail from .* with length %d", alteredValue));
+
+        dropTable("test_iceberg_file_size");
+    }
+
+    private static HdfsEnvironment createTestHdfsEnvironment()
```

Comment:
Use `HDFS_ENVIRONMENT` from `HiveTestUtils`. You only need to create one yourself if you need different config or have other special requirements.